st2actions/st2actions/query/base.py

```diff
@@ -0,0 +1,150 @@
+
+    def _query_and_save_results(self, query_context):
+        execution_id = query_context.execution_id
+        actual_query_context = query_context.query_context
+
+        LOG.debug('Querying external service for results. Context: %s' % actual_query_context)
+        try:
+            (status, results) = self.query(execution_id, actual_query_context)
+        except:
+            LOG.exception('Failed querying results for action_execution_id %s.', execution_id)
+            self._delete_state_object(query_context)
+            return
+
+        done = (status in DONE_STATES)
+
+        try:
+            self._update_action_results(execution_id, status, results)
+        except Exception:
+            LOG.exception('Failed updating action results for action_execution_id %s',
+                          execution_id)
+            self._delete_state_object(query_context)
+            return
+        else:
+            if done:
+                self._delete_state_object(query_context)
+                return
+
+        self._query_contexts.put((time.time(), query_context))
+        return
+
+    def _update_action_results(self, execution_id, status, results):
+        actionexec_db = ActionExecution.get_by_id(execution_id)
+        if not actionexec_db:
+            raise Exception('No DB model for action_execution_id: %s' % execution_id)
+        actionexec_db.result = results
+        actionexec_db.status = status
+        updated_exec = ActionExecution.add_or_update(actionexec_db)
+        return updated_exec
+
+    def _delete_state_object(self, query_context):
+        state_db = ActionExecutionState.get_by_id(query_context.id)
+        if state_db is not None:
+            try:
+                ActionExecutionState.delete(state_db)
+            except:
+                LOG.exception('Failed clearing state object: %s', state_db)
+
+    def query(self, execution_id, query_context):
+        """
+        This is the method individual queriers must implement.
+        This method should return a tuple of (status, results).
+
+        status should be one of ACTIONEXEC_STATUS_SUCCEEDED, ACTIONEXEC_STATUS_RUNNING,
+        ACTIONEXEC_STATUS_FAILED defined in st2common.constants.action.
+        """
+        pass
+
+    def print_stats(self):
+        LOG.info('\t --- Name: %s, pending queuries: %d', self.__class__.__name__,
+                 self._query_contexts.qsize())
+
+
+class QueryContext(object):
+    def __init__(self, obj_id, execution_id, query_context, query_module):
+        self.id = obj_id
+        self.execution_id = execution_id
+        self.query_context = query_context
+        self.query_module = query_module
+
+    @classmethod
+    def from_model(cls, model):
+        return QueryContext(str(model.id), str(model.execution_id), model.query_context,
+                            model.query_module)
+
+    def __repr__(self):
+        return ('<QueryContext id=%s,execution_id=%s,query_context=%s>' %
```

Comment:
I used this from one of our existing reprs. If you point me to something you like, I can fix this :)